src/Modal.php

```diff
@@ -33,17 +35,21 @@ public function basePageRoute(string $name, mixed $parameters = [], bool $absolu
         return $this->baseRoute($name, $parameters, $absolute);
     }
 
-    public function with(array $props): static
+    public function with(array|Arrayable $props): static
     {
-        $this->props = $props;
+        $this->props = $props instanceof Arrayable ? $props->toArray() : $props;
 
         return $this;
     }
 
     public function render(): mixed
     {
+        $flatProps = [];
+        foreach ($this->props as $key => $prop) {
+            $flatProps['modal.props.'.$key] = $prop;
+        }
         /** @phpstan-ignore-next-line */
-        inertia()->share(['modal' => $this->component()]);
+        inertia()->share(['modal' => $this->component(), ...$flatProps]);
```

Comment:
Manually looping over the props can be avoided by using the `Arr::dot` [helper method](https://laravel.com/docs/11.x/helpers#method-array-dot):

```php
inertia()->share(['modal' => $this->component(), ...Arr::dot($props, 'modal.props.')]);
```